nexus/types/src/deployment.rs

```diff
@@ -271,31 +295,159 @@ impl<'a> fmt::Display for BlueprintDisplay<'a> {
+            // `OmicronZonesConfig`.
+            generation: collection.generation,
+            zones,
+        };
+        ret.sort();
+
+        ret
+    }
+
+    /// Sorts the list of zones stored in this configuration.
+    ///
+    /// This is not strictly necessary. But for testing, it's helpful for
+    /// things to be in sorted order.
+    pub fn sort(&mut self) {
+        self.zones.sort_unstable_by_key(|z| z.config.id);
+    }
+
+    /// Converts self to an [`OmicronZonesConfig`].
+    pub fn to_omicron_zones_config(&self) -> OmicronZonesConfig {
+        OmicronZonesConfig {
+            generation: self.generation,
+            zones: self.zones.iter().map(|z| z.config.clone()).collect(),
+        }
+    }
+}
+
+/// Describes one Omicron-managed zone in a blueprint.
+///
+/// This is a wrapper around an [`OmicronZoneConfig`] that also includes a
+/// [`BlueprintZonePolicy`].
+///
+/// Part of [`BlueprintZonesConfig`].
+#[derive(Debug, Clone, Eq, PartialEq, JsonSchema, Deserialize, Serialize)]
+pub struct BlueprintZoneConfig {
+    /// The underlying zone configuration.
+    pub config: OmicronZoneConfig,
+
+    /// The policy for this zone.
+    pub zone_policy: BlueprintZonePolicy,
+}
+
+impl BlueprintZoneConfig {
+    /// Return a struct that can be displayed to present information about the
+    /// zone.
+    pub fn display(&self) -> BlueprintZoneConfigDisplay<'_> {
+        BlueprintZoneConfigDisplay { zone: self }
+    }
+}
+
+/// A wrapper to allow a [`BlueprintZoneConfig`] to be displayed with
+/// information.
+///
+/// Returned by [`BlueprintZoneConfig::display()`].
+#[derive(Clone, Debug)]
+#[must_use = "this struct does nothing unless displayed"]
+pub struct BlueprintZoneConfigDisplay<'a> {
+    zone: &'a BlueprintZoneConfig,
+}
+
+impl<'a> fmt::Display for BlueprintZoneConfigDisplay<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let z = self.zone;
+        write!(
+            f,
+            "{} {:<width$} {} [underlay IP {}]",
+            z.config.id,
+            z.zone_policy,
+            z.config.zone_type.label(),
+            z.config.underlay_address,
+            width = BlueprintZonePolicy::DISPLAY_WIDTH,
+        )
+    }
+}
+
+/// The policy for an Omicron-managed zone in a blueprint.
+///
+/// Part of [`BlueprintZoneConfig`].
+#[derive(
+    Debug, Copy, Clone, Eq, PartialEq, JsonSchema, Deserialize, Serialize,
+)]
+#[serde(rename_all = "snake_case")]
+pub enum BlueprintZonePolicy {
+    InService,
+    NotInService,
+}
+
+impl BlueprintZonePolicy {
+    /// The maximum width of `Display` output.
+    const DISPLAY_WIDTH: usize = 14;
+}
+
+impl fmt::Display for BlueprintZonePolicy {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            // Neither `write!(f, "...")` nor `f.write_str("...")` obey fill
+            // and alignment (used above), but this does.
```

Comment:
What! So writing a correct Display impl requires knowing whether your callers will use width and alignment modifiers?

Reply:
TIL: yes, yes it does.

Reply:
:)